cql3/query_processor.cc

```diff
@@ -794,6 +794,49 @@ query_options query_processor::make_internal_options(
     return query_options(cl, bound_values);
 }
 
+query_options query_processor::make_internal_options(
+        const statements::prepared_statement::checked_weak_ptr& p,
+        const std::vector<std::optional<data_value>>& opt_values,
```

Comment:
We shouldn't use optional for unset. optional is very common, and is already used for null. Will a reader seeing std::nullopt know the meaning here is unset and not null? Probably not.

That's why I suggested variant<data_value, unset_value>. An alternative is to introduce maybe_unset_data_value that wraps optional<data_value>, but has explict ways to construct unsets (and implicit ways to construct set data_values).

Reply:
Also,  why is this a new overload?

Reply:
> We shouldn't use optional for unset. optional is very common, and is already used for null.

Used for null `data_value`?  I though that's what `data_value::make_null()` is for.


> Will a reader seeing std::nullopt know the meaning here is unset and not null? Probably not.
> 
> That's why I suggested variant<data_value, unset_value>. An alternative is to introduce maybe_unset_data_value that wraps optional<data_value>, but has explict ways to construct unsets (and implicit ways to construct set data_values).

My first attempt (that I didn't publish) was to extend data_value to hold an unset value in addition to a null value, but that was overly complicated and intrusive.

I realized that a null value is considered a data_value while a unset value is not a value.  It's a lack of thereof.
That's why a disengaged `optional<data_value>` should mean "no value" rather than "null value".

But I can live with the variant as well.

> Also, why is this a new overload?

To save an extra allocation to store the initializer_list into a vector in the common case.

Reply:
> > We shouldn't use optional for unset. optional is very common, and is already used for null.
> 
> Used for null `data_value`? I though that's what `data_value::make_null()` is for.


Not for null data_value, but for null managed_bytes_opt. At the caller site it's easy to mistake one of the other.

In any case std::nullopt doesn't hint at UNSET, which is an obscure feature, so no one will think of it.

> 
> > Will a reader seeing std::nullopt know the meaning here is unset and not null? Probably not.
> > That's why I suggested variant<data_value, unset_value>. An alternative is to introduce maybe_unset_data_value that wraps optional<data_value>, but has explict ways to construct unsets (and implicit ways to construct set data_values).
> 
> My first attempt (that I didn't publish) was to extend data_value to hold an unset value in addition to a null value, but that was overly complicated and intrusive.


Yes. It's also inserting an obscure feature into a main API.

> 
> I realized that a null value is considered a data_value while a unset value is not a value. It's a lack of thereof. That's why a disengaged `optional<data_value>` should mean "no value" rather than "null value".


It's too subtle. People are used to seeing std::nullopt as null due to managed_bytes_opt.

> 
> But I can live with the variant as well.
> 
> > Also, why is this a new overload?
> 
> To save an extra allocation to store the initializer_list into a vector in the common case.


All of this is a slow path for metadata. The data path builds mutations directly.

initializer_list<> is a huge pessimiziation since it can't be moved from.